Consume in reverese frozen mutations

zdzblo-trawy opened the pull request:
This series introduces consuming frozen mutations in reverse.  It also strips the `to_data_query_result` from unfrozing the mutation before being consumed in reverse, which is a time-consuming operation. 
Only `consume_in_reverse::yes` is supported, as the `consume_in_reverse::legacy_half_reverse` support is not needed anywhere.

Fixes #11199

zdzblo-trawy commented:
CC @bhalevy

bhalevy commented on frozen_mutation.hh:
```diff
@@ -301,7 +301,8 @@ auto frozen_mutation::consume(schema_ptr s, frozen_mutation_consumer_adaptor<Con
 template<FlattenedConsumerV2 Consumer>
 auto frozen_mutation::consume(schema_ptr s, Consumer& consumer, consume_in_reverse reverse) const -> frozen_mutation_consume_result<decltype(consumer.consume_end_of_stream())> {
-    frozen_mutation_consumer_adaptor adaptor(s, consumer);
+    schema_ptr ss = reverse == consume_in_reverse::yes ? s->make_reversed() : s;
```
can we make consume_in_reverse a template parameter so the condition can be evaluated at compile time?

zdzblo-trawy replied:
I do not think it matters here. This is not a super hot code and making it a compile time parameter will increase binary size (and this is probably the single worst CPU performance issue Scylla currently has). It also doesn't quite matter, as it is very probable that the compiler already inlines the code and notices that the value of `reverse` in parent function (say, `to_data_query_result` doesn't change so it can be taken out of the loop. And also such a change would mean that I need to increase code bloat in `to_data_query_result` that would need effectively to be copy-pasted.

avikivity replied:
Agree. If the reverse parameter ends up trickling down to add branches everywhere, then it's important, but I don't think that's the case.

bhalevy commented on mutation_partition_view.cc:
```diff
@@ -286,6 +286,7 @@ mutation_partition_view::accept_ordered_cookie<reverse>::rts_crs_iterators::rts_
 
 template <bool is_preemptible, consume_in_reverse reverse>
 mutation_partition_view::accept_ordered_result<reverse> mutation_partition_view::do_accept_ordered(const schema& schema, mutation_partition_view_virtual_visitor& visitor, accept_ordered_cookie<reverse> cookie) const {
+    static_assert(reverse != consume_in_reverse::legacy_half_reverse);
```
The patches that get rid of legacy_half_reverse should come first, potentially even merged separately, if this PR gets delayed for some reason.
This will simplify the unit tests that are added early on, as they are bogus for legacy_half_revrerse

zdzblo-trawy replied:
I hoped to avoid that (it seems like a fair bit of useless work) but if you deem that necessary I can do that.

bhalevy replied:
yes, please, we try to get the patches in the best shape we can

bhalevy commented on test/boost/mutation_query_test.cc:
```diff
@@ -564,6 +564,70 @@ SEASTAR_THREAD_TEST_CASE(test_result_size_calculation) {
+    };
+
+    BOOST_REQUIRE_EQUAL(res_frozen.result.size(), res_unfrozen.result.size());
+    for (auto& res : res_frozen.result) {
```
nit: const auto&

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3548/

avikivity commented:
Please adjust the cover letter to explain why we're doing this (not just what). What's the benefit?

zdzblo-trawy commented:
> Please adjust the cover letter to explain why we're doing this (not just what). What's the benefit?

I slightly reworded the cover letter so that it emphasizes more the expected gains.

avikivity commented:
I guess avoiding the time-consuming operation is the goal, but it's probably an unrealistic expectation to have that spelled out.

avikivity commented on mutation_partition_view.cc:
```diff
@@ -273,7 +274,7 @@ future<> mutation_partition_view::do_accept_gently(const column_mapping& cm, Vis
 }
 
-template <bool is_preemptible>
+template <bool is_preemptible, consume_in_reverse reverse>
```
I won't ask to redo the series, but splitting it by first adding the parameters everywhere, then adding the contents, isn't helping review. For example, how does the reviewer know you haven't forgotten an implementation somewhere?

Patches shouldn't add bugs, and this one does: an API that accepts an instruction to reverse, then ignores it.

zdzblo-trawy replied:
I will at least add a comment explaining it.

avikivity commented on serializer_impl.hh:
```diff
@@ -851,6 +860,23 @@ public:
+    template<bool B> vector_deserializer<T, B> to_forward() &&;
+    template<>
+    vector_deserializer<T, true> to_forward<true>() && {
+        static_assert(IsForward, "cannot convert reverse vector deserializer into forward one");
```
Note: nicer to have `requires IsForward` as a static_assert implies something wrong with the callee, not the caller.

avikivity replied:
Also: does it have to be a template? I guess it does with static_assert to prevent early assertion.

avikivity replied:
I'm sure it can be improved. But first we need to agree if we want the thing or not.

avikivity commented:
The series looks good, never mind my complaints. But I have a question about the approach: should we rather send the mutations in reverse? The replica already generates them in reversed form, then we unreverse them to send in forward form, then we reverse them again (or consume in reverse as this series does).

It would be much simpler to send reversed frozen mutations, and consume them in the forwards direction, no?

It requires new verbs, but ends up simplifying things.

bhalevy commented:
> The series looks good, never mind my complaints. But I have a question about the approach: should we rather send the mutations in reverse? The replica already generates them in reversed form, then we unreverse them to send in forward form, then we reverse them again (or consume in reverse as this series does).
> 
> It would be much simpler to send reversed frozen mutations, and consume them in the forwards direction, no?
> 
> It requires new verbs, but ends up simplifying things.

that's pretty much the plan for the future.
the gist of it is to implement "native" reverse query option (to be called reverse_v2) that will replace the legacy reverse query option (modulo a cluster feature bit).

In this mode the replica will send the result already reversed and the coordinator can then consume it in the natural order using the reversed schema.

That said, I anticipate we may still want to have the consume in reverse mode, at least for the transition for a couple releases.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3562/

avikivity commented:
> > The series looks good, never mind my complaints. But I have a question about the approach: should we rather send the mutations in reverse? The replica already generates them in reversed form, then we unreverse them to send in forward form, then we reverse them again (or consume in reverse as this series does).
> > It would be much simpler to send reversed frozen mutations, and consume them in the forwards direction, no?
> > It requires new verbs, but ends up simplifying things.
> 
> that's pretty much the plan for the future. the gist of it is to implement "native" reverse query option (to be called reverse_v2) that will replace the legacy reverse query option (modulo a cluster feature bit).
> 
> In this mode the replica will send the result already reversed and the coordinator can then consume it in the natural order using the reversed schema.
> 
> That said, I anticipate we may still want to have the consume in reverse mode, at least for the transition for a couple releases.

But we already have consume in reverse, it's just inefficient.

Isn't it better to spend the effort on send-in-reverse?

bhalevy commented:
> > > The series looks good, never mind my complaints. But I have a question about the approach: should we rather send the mutations in reverse? The replica already generates them in reversed form, then we unreverse them to send in forward form, then we reverse them again (or consume in reverse as this series does).
> > > It would be much simpler to send reversed frozen mutations, and consume them in the forwards direction, no?
> > > It requires new verbs, but ends up simplifying things.
> > 
> > 
> > that's pretty much the plan for the future. the gist of it is to implement "native" reverse query option (to be called reverse_v2) that will replace the legacy reverse query option (modulo a cluster feature bit).
> > In this mode the replica will send the result already reversed and the coordinator can then consume it in the natural order using the reversed schema.
> > That said, I anticipate we may still want to have the consume in reverse mode, at least for the transition for a couple releases.
> 
> But we already have consume in reverse, it's just inefficient.
> 
> Isn't it better to spend the effort on send-in-reverse?

The effort was already spent.  It'd be a shame not to use it because there's a better solution in the future that requires more work, while we have an opportunity to improve performance with a simple low-risk change.

We realized the potential for sending in reverse while reviewing these patches internally.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3563/

avikivity commented:
> > > > The series looks good, never mind my complaints. But I have a question about the approach: should we rather send the mutations in reverse? The replica already generates them in reversed form, then we unreverse them to send in forward form, then we reverse them again (or consume in reverse as this series does).
> > > > It would be much simpler to send reversed frozen mutations, and consume them in the forwards direction, no?
> > > > It requires new verbs, but ends up simplifying things.
> > > 
> > > 
> > > that's pretty much the plan for the future. the gist of it is to implement "native" reverse query option (to be called reverse_v2) that will replace the legacy reverse query option (modulo a cluster feature bit).
> > > In this mode the replica will send the result already reversed and the coordinator can then consume it in the natural order using the reversed schema.
> > > That said, I anticipate we may still want to have the consume in reverse mode, at least for the transition for a couple releases.
> > 
> > 
> > But we already have consume in reverse, it's just inefficient.
> > Isn't it better to spend the effort on send-in-reverse?
> 
> The effort was already spent. It'd be a shame not to use it because there's a better solution in the future that requires more work, while we have an opportunity to improve performance with a simple low-risk change.

There is always risk. This is a less-tested area since reversed queries are rare. 

> 
> We realized the potential for sending in reverse while reviewing these patches internally.

bhalevy commented:
@zdzblo-trawy lets start with extracting out the patches to get rid of consume_in_reverse::legacy_half_reverse they can be discussed and be merged independently of consuming frozen_mutation in reverse.

zdzblo-trawy commented:
Last forced push: Reordered commits, removed useless tests for consuming in legacy_half_reverse, addressed the complaint about unused `reverse` parameter.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3610/

bhalevy commented:
> Last forced push: Reordered commits, removed useless tests for consuming in legacy_half_reverse, addressed the complaint about unused `reverse` parameter. Hopefully now it is mergable.

@zdzblo-trawy, can you please submit the patches to get rid of consume_in_reverse::legacy_half_reverse (and probably turn it into an enum class) as a separate PR?

avikivity commented:
> > Last forced push: Reordered commits, removed useless tests for consuming in legacy_half_reverse, addressed the complaint about unused `reverse` parameter. Hopefully now it is mergable.
> 
> @zdzblo-trawy, can you please submit the patches to get rid of consume_in_reverse::legacy_half_reverse (and probably turn it into an enum class) as a separate PR?

What's the goal?

bhalevy commented:
> > > Last forced push: Reordered commits, removed useless tests for consuming in legacy_half_reverse, addressed the complaint about unused `reverse` parameter. Hopefully now it is mergable.
> > 
> > 
> > @zdzblo-trawy, can you please submit the patches to get rid of consume_in_reverse::legacy_half_reverse (and probably turn it into an enum class) as a separate PR?
> 
> What's the goal?

consume_in_reverse::legacy_half_reverse mode was defined to produce a reversed stream where range tombstones were ordered (in reverse) by their end bound, but each range tombstone was not reversed, keeping its start bound <= end bound in forward order.

Now with mutation_fragment_v2 consuming a mutation produces only range_tombstone_change fragments,
while no range_tombstone fragments are produced anymore, therefore legacy_half_reverse mode is not applicable.

According to @zdzblo-trawy's tests consuming in consume_in_reverse::legacy_half_reverse mode produces an invalid mutation fragments "stream" where all clustering rows are emitted first (in reversed order), then all range_tombstone_change:s are emitted (IIRC in reverse order as well).  This works in practice since we don't read as a mutation_fragments stream, but rather it is use to build the query result.

Instead, we should currently use the consume_in_reverse::yes mode and get rid of the defunct legacy_half_reverse mode.

avikivity commented:
> > > > Last forced push: Reordered commits, removed useless tests for consuming in legacy_half_reverse, addressed the complaint about unused `reverse` parameter. Hopefully now it is mergable.
> > > 
> > > 
> > > @zdzblo-trawy, can you please submit the patches to get rid of consume_in_reverse::legacy_half_reverse (and probably turn it into an enum class) as a separate PR?
> > 
> > 
> > What's the goal?
> 
> consume_in_reverse::legacy_half_reverse mode was defined to produce a reversed stream where range tombstones were ordered (in reverse) by their end bound, but each range tombstone was not reversed, keeping its start bound <= end bound in forward order.
> 
> Now with mutation_fragment_v2 consuming a mutation produces only range_tombstone_change fragments, while no range_tombstone fragments are produced anymore, therefore legacy_half_reverse mode is not applicable.
> 
> According to @zdzblo-trawy's tests consuming in consume_in_reverse::legacy_half_reverse mode produces an invalid mutation fragments "stream" where all clustering rows are emitted first (in reversed order), then all range_tombstone_change:s are emitted (IIRC in reverse order as well). This works in practice since we don't read as a mutation_fragments stream, but rather it is use to build the query result.
> 
> Instead, we should currently use the consume_in_reverse::yes mode and get rid of the defunct legacy_half_reverse mode.

Sounds reasonable.

denesb commented:
> > The effort was already spent. It'd be a shame not to use it because there's a better solution in the future that requires more work, while we have an opportunity to improve performance with a simple low-risk change.
> 
> There is always risk. This is a less-tested area since reversed queries are rare.

I also have my doubts about the usefulness of consuming frozen mutations in reverse: the mutations we'd be consuming in reverse are produced by internal code so they can be expected to be reasonably small, thus the cost of reversing after unfreezing is bounded. It feels bad to throw away already spent effort, but like @avikivity said, this is not a well stressed code path and the cost/benefit/risk math may not work out.

zdzblo-trawy commented:
@bhalevy this PR now solely treats consuming frozen mutations in reverse

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3760/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3761/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3762/

bhalevy commented:
https://github.com/scylladb/scylladb/issues/12557 was opened to track native reverse queries
